cluster/kubernetes/namespacer.go

```diff
@@ -0,0 +1,166 @@
+		return "", nil
+	}
+	return m.GetNamespace(), nil
+}
+
+func (n *namespaceViaDiscovery) lookupNamespaced(groupVersion, kind string) (bool, error) {
+	resourceList, err := n.disco.ServerResourcesForGroupVersion(groupVersion)
+	if err != nil {
+		return false, fmt.Errorf("error looking up API resources for %s.%s: %s", kind, groupVersion, err.Error())
+	}
+	for _, resource := range resourceList.APIResources {
+		if resource.Kind == kind {
+			return resource.Namespaced, nil
+		}
+	}
+	return false, fmt.Errorf("resource not found for API %s, kind %s", groupVersion, kind)
+}
+
+// This mainly exists so I can put the controller and store in here,
+// so as to have access to them for testing; and, so that we can do
+// our own invalidation.
+type cachedDiscovery struct {
+	discovery.CachedDiscoveryInterface
+
+	store      toolscache.Store
+	controller toolscache.Controller
+
+	invalidMu sync.Mutex
+	invalid   bool
+}
+
+// The older (v8.0.0) implementation of MemCacheDiscovery refreshes
+// the cached values, synchronously, when Invalidate is called. Since
+// we will invalidate every time something cahnges, but it only
+// matters when we want to read the cached values, this method (and
+// ServerResourcesForGroupVersion) saves the invalidation for when a
+// read is done.
+func (d *cachedDiscovery) Invalidate() {
+	d.invalidMu.Lock()
+	d.invalid = true
+	d.invalidMu.Unlock()
+}
+
+// This happens to be the method that we call in the namespacer; so,
+// this is the one where we check whether the cache has been
+// invalidated. A cachedDiscovery implementation for more general use
+// would do this for all methods (that weren't implemented purely in
+// terms of other methods).
+func (d *cachedDiscovery) ServerResourcesForGroupVersion(groupVersion string) (*metav1.APIResourceList, error) {
+	d.invalidMu.Lock()
+	invalid := d.invalid
+	d.invalid = false
+	d.invalidMu.Unlock()
+	if invalid {
+		d.CachedDiscoveryInterface.Invalidate()
+	}
+	return d.CachedDiscoveryInterface.ServerResourcesForGroupVersion(groupVersion)
+}
+
+type chainHandler struct {
+	first toolscache.ResourceEventHandler
+	next  toolscache.ResourceEventHandler
+}
+
+func (h chainHandler) OnAdd(obj interface{}) {
+	h.first.OnAdd(obj)
+	h.next.OnAdd(obj)
+}
+
+func (h chainHandler) OnUpdate(old, new interface{}) {
+	h.first.OnUpdate(old, new)
+	h.next.OnUpdate(old, new)
+}
+
+func (h chainHandler) OnDelete(old interface{}) {
+	h.first.OnDelete(old)
+	h.next.OnDelete(old)
+}
+
+// MakeCachedDiscovery constructs a CachedDicoveryInterface that will
+// be invalidated whenever the set of CRDs change. The idea is that
+// the only avenue of a change to the API resources in a running
+// system is CRDs being added, updated or deleted. The prehandlers are
+// there to allow us to put extra synchronisation in, for testing.
+func MakeCachedDiscovery(d discovery.DiscoveryInterface, c crd.Interface, shutdown <-chan struct{}, prehandlers ...toolscache.ResourceEventHandler) *cachedDiscovery {
```

Comment:
The `cachedDiscovery` is only used within the `namespaceViaDiscovery`, but I understand you want to reusecachedDiscovery` for the garbage collection during syncing, and thus it makes sense to pass it as a parameter instead of creating in internally (requiring to make the constructor public). 

However, in that case, I think it makes sense to put the `cachedDiscovery` implementation in a separate file.